lib/recursive-methods.rb

```diff
@@ -1,49 +1,130 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O of n because the program must perform as many operations as n. 
+# Space complexity: O of n because a new function is placed on the stack for each operation
+
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  if n < 0
+    raise ArgumentError, 'Please enter a positive number' 
+  elsif n == 1 || n == 0
+    return 1
+  else
+    return n * factorial(n - 1)
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: Since the recursive step is acting on two elements at a time, it's O of n/2 for even string lengths and O of n/2 + 1 for odd lengths where n is the integer length of the string.
+# Space complexity: Same as time complexity. 
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  # nil case and base case
+  if s.length == 0 || s.length == 1
+    return s
+  else 
+    # recursive step
+    return s[-1] + reverse(s[1..-2]) + s[0]
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: Same as above -- O of n/2 for even string lengths and O of n/2 + 1 where n is the length of the string.
+# Space complexity: Same as time complexity.
+# Note: I think this is reversing in place because we're not creating any new variables?
 def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+  # nil case and base case
+  if s.length == 0 || s.length == 1
+    return s
+  else 
+    # recursive step
+    return s[-1] + reverse_inplace(s[1..-2]) + s[0]
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O of n where n is the number of bunnies :)
+# Space complexity: O of n
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+  if n == 0
+    return 0
+  elsif n == 1
+    return 2
+  else 
+    return 2 + bunny(n-1)
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O of n/2 where n is an even number of parens in the string, O of n/2 + 1 when the string character count is odd. 
+# Space complexity: Same as time complexity. 
 def nested(s)
-    raise NotImplementedError, "Method not implemented"
+  if s.length == 0
+    return true
+  elsif s.length == 1
+    return false
+  elsif s.length == 2 && s[0] == ")" || s.length == 2 && s[-1] == "("
+    return false
+  elsif s.length == 2 && s[0] == "(" && s[-1] == ")"
+    return true
+  else
+    s.slice!(0)
+    s.slice!(-1)
+    return nested(s)
+  end
+  return false
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O of n where n is the number of elements in the array
+# Space complexity: O of n
 def search(array, value)
```

Comment:
Because you are deleting elements in the array with each iteration, this has time/space complexities O(n<sup>2</sup>)